File: queryoptimiser/Typescript/src/seed.ts
// ------------------------
// 1. Import dependencies
// ------------------------
import { faker } from "@faker-js/faker";   // Faker for generating fake data
import sequelize from "./db";               // Sequelize instance
import User from "./models/User";           // User model
import Order from "./models/Order";         // Order model
import OrderItem from "./models/OrderItem"; // OrderItem model

// ------------------------
// 2. Seed function
// ------------------------
async function seed() {
  // Sync all models with the database
  // `force: true` drops existing tables and recreates them
  await sequelize.sync({ force: true });

  console.log("Database synced. Seeding data...");

  // ------------------------
  // Constants
  // ------------------------
  const NUM_USERS = 2000;
  const ORDERS_PER_USER = 10;
  const ITEMS_PER_ORDER = 5;

  // ------------------------
  // Step 1: Generate all users
  // ------------------------
  const usersData = Array.from({ length: NUM_USERS }, () => ({
    name: faker.name.fullName()
  }));

  // Bulk insert users and get inserted rows with IDs
  const users = await User.bulkCreate(usersData, { returning: true });

  // ------------------------
  // Step 2: Generate all orders
  // ------------------------
  const ordersData: { userId: number }[] = [];
  for (const user of users) {
    for (let j = 0; j < ORDERS_PER_USER; j++) {
      ordersData.push({ userId: user.id });
    }
  }

  // Bulk insert orders and get inserted rows with IDs
  const orders = await Order.bulkCreate(ordersData, { returning: true });

  // ------------------------
  // Step 3: Generate all order items
  // ------------------------
  const itemsData: { orderId: number; productName: string }[] = [];
  for (const order of orders) {
    for (let k = 0; k < ITEMS_PER_ORDER; k++) {
      itemsData.push({
        orderId: order.id,
        productName: faker.commerce.productName()
      });
    }
  }

  // Bulk insert all items
  await OrderItem.bulkCreate(itemsData);

  console.log("Database seeded successfully!");
  process.exit(0); // Exit process after seeding
}

// ------------------------
// 3. Run the seed function
// ------------------------
seed().catch((err) => {
  console.error("Seeding failed:", err);
  process.exit(1);
});

File: queryoptimiser/python/seed.py
from faker import Faker               # Import Faker to generate fake data
from database import SessionLocal     # Import session factory to interact with the DB
from models import User, Order, OrderItem  # Import ORM models (tables)

# Initialize Faker to generate random names, words, etc.
fake = Faker()

# Create a new database session
db = SessionLocal()

# Define how much fake data we want to generate
NUM_USERS = 2000          # Number of users to create
ORDERS_PER_USER = 10      # Number of orders per user
ITEMS_PER_ORDER = 5       # Number of items per order

# Loop to create users
for _ in range(NUM_USERS):
    # Create a new User instance with a random name
    user = User(name=fake.name())
    db.add(user)           # Add the user to the session (not yet committed)
    db.flush()             # Flush to the DB to assign an ID to the user
                           # (needed for foreign key in orders)

    # Loop to create orders for this user
    for _ in range(ORDERS_PER_USER):
        order = Order(user_id=user.id)  # Associate order with current user
        db.add(order)                   # Add order to session
        db.flush()                      # Flush to get order.id for items

        # Loop to create items for this order
        for _ in range(ITEMS_PER_ORDER):
            item = OrderItem(
                order_id=order.id,        # Associate item with current order
                product_name=fake.word()  # Generate a random product name
            )
            db.add(item)                  # Add item to session

# Commit all changes to the database
# This writes all users, orders, and items permanently
db.commit()

# Close the session to free up resources
db.close()

# Print confirmation message
print("Database seeded.")
